include/dlaf/permutations/general/impl.h

```diff
@@ -152,4 +157,345 @@ void Permutations<B, D, T, C>::call(SizeType i_begin, SizeType i_end, Matrix<con
+  return ex::ensure_started(
+      di::transform<di::TransformDispatchType::Plain, false>(di::Policy<DefaultBackend_v<D>>{},
+                                                             std::move(counts_fn), std::move(sender)));
+}
+
+template <Device D, Coord C>
+void copyLocalPartsFromGlobalIndex(SizeType i_loc_begin, const matrix::Distribution& dist,
+                                   Matrix<const SizeType, D>& global_index,
+                                   Matrix<SizeType, D>& local_index) {
+  namespace ex = pika::execution::experimental;
+
+  for (auto tile_wrt_local : common::iterate_range2d(local_index.distribution().localNrTiles())) {
+    SizeType i_gl_tile = dist.globalTileFromLocalTile<C>(i_loc_begin + tile_wrt_local.row());
+    ex::start_detached(ex::when_all(global_index.read_sender(GlobalTileIndex(i_gl_tile, 0)),
+                                    local_index.readwrite_sender(tile_wrt_local)) |
+                       dlaf::matrix::copy(dlaf::internal::Policy<DefaultBackend_v<D>>{}));
+  }
+}
+
+// @param index_map a column matrix that represents a map from local `out` to local `in` indices
+template <class T, Device D, Coord C, class IndexMapSender, class InSender, class OutSender>
+void applyPackingIndex(const matrix::Distribution& subm_dist, IndexMapSender&& index_map, InSender&& in,
+                       OutSender&& out) {
+  namespace ex = pika::execution::experimental;
+  namespace di = dlaf::internal;
+
+  auto sender = ex::when_all(std::forward<IndexMapSender>(index_map), std::forward<InSender>(in),
+                             std::forward<OutSender>(out));
+
+  auto permute_fn = [subm_dist](const auto& index_tile_futs, const auto& mat_in_tiles,
+                                const auto& mat_out_tiles, auto&&... ts) {
+    const SizeType* i_ptr = index_tile_futs[0].get().ptr();
+    applyPermutations<T, D, C>(GlobalElementIndex(0, 0), subm_dist.size(), 0, subm_dist, i_ptr,
+                               mat_in_tiles, mat_out_tiles, std::forward<decltype(ts)>(ts)...);
+  };
+  ex::start_detached(
+      di::transform<di::TransformDispatchType::Plain, false>(di::Policy<DefaultBackend_v<D>>(),
+                                                             std::move(permute_fn), std::move(sender)));
+}
+
+template <Device D, class InTileSender, class OutTileSender>
+void transposeTileSenders(InTileSender&& in, OutTileSender&& out) {
+  namespace ex = pika::execution::experimental;
+  namespace di = dlaf::internal;
+
+  auto sender = ex::when_all(std::forward<InTileSender>(in), std::forward<OutTileSender>(out));
+
+  auto transpose_fn = [](const auto& in_tile, const auto& out_tile) {
+    for (TileElementIndex idx : common::iterate_range2d(out_tile.size())) {
+      out_tile(idx) = in_tile(transposed(idx));
+    }
+  };
+
+  ex::start_detached(
+      di::transform(di::Policy<DefaultBackend_v<D>>(), std::move(transpose_fn), std::move(sender)));
```

Comment:
Same comment here as in `initPackingIndex`

Reply:
https://github.com/eth-cscs/DLA-Future/pull/676/commits/1150fae0afcb62d9b217f3feaf67ba07c3fede9d